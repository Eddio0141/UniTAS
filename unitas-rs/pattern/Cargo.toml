[package]
name = "pattern"
version = "0.1.0"
edition = "2024"

[dependencies]
