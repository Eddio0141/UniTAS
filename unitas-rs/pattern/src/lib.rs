#[cfg(test)]
mod tests;

#[derive(Debug, PartialEq, Eq)]
pub struct Pattern {
    pattern: &'static [Option<u8>],
    offset: isize,
}

impl Pattern {
    pub const fn new(offset: isize, pattern: &'static [Option<u8>]) -> Self {
        if pattern.is_empty() {
            panic!("pattern must not be empty");
        }

        if pattern[0].is_none() {
            panic!("pattern must not start with None");
        }

        if pattern[pattern.len() - 1].is_none() {
            panic!("pattern must not end with None");
        }

        Self { pattern, offset }
    }

    /// Tries to match the pattern against the data.
    /// - If the pattern matches, returns `Some(offset)`.
    /// - If the pattern doesn't match, returns `None`.
    /// - The offset would be the offset of the first byte of the pattern.
    pub fn matches(&self, mut data: &[u8]) -> Option<usize> {
        // we don't let pattern to be constructed by the user, so head is always Some
        let head = self.pattern[0].unwrap();
        let mut data_offset = 0;

        'data_loop: loop {
            // advance data slice to head
            data = match data.iter().position(|&f| f == head) {
                Some(index) => {
                    data_offset += index;
                    &data[index..]
                }
                None => return None,
            };

            // not enough bytes?
            if data.len() < self.pattern.len() {
                return None;
            }

            // check if pattern matches
            for (i, pattern) in self.pattern[1..].iter().enumerate() {
                if let Some(pattern) = pattern {
                    if data[i + 1] != *pattern {
                        data_offset += 1;
                        data = &data[1..];
                        continue 'data_loop;
                    }
                }
            }

            let offset = data_offset.checked_add_signed(self.offset)?;

            return Some(offset);
        }
    }
}
