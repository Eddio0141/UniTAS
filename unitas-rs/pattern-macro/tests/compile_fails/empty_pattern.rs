use pattern_macro::pattern;

fn main() {
    pattern!(0, "");
}
