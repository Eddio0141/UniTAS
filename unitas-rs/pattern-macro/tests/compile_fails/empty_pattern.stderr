error: what's the point of an empty pattern?
 --> tests/compile_fails/empty_pattern.rs:4:17
  |
4 |     pattern!(0, "");
  |                 ^^
